titus-server-master/src/main/java/com/netflix/titus/master/scheduler/constraint/TaskCache.java

```diff
@@ -53,36 +58,80 @@ public void prepare() {
         return currentCacheValue.get().getTasksByZoneIdCounters(jobId);
     }
 
+    // Returns a task ID if there is a task assigned to the provided IP allocation
+    public Optional<String> getTaskByIpAllocationId(String ipAllocationId) {
+        return Optional.ofNullable(currentCacheValue.get().assignedIpAllocations.get(ipAllocationId));
+    }
+
+    // Updates the cache to reflect assignment of an IP allocation to a task
+    public void addTaskIpAllocation(String ipAllocationId, String taskId) {
+        currentCacheValue.get().assignedIpAllocations.put(ipAllocationId, taskId);
+    }
+
+    public Optional<String> getZoneIdByIpAllocationId(String ipAllocationId) {
+        return Optional.ofNullable(currentCacheValue.get().ipAllocationIdToZoneId.get(ipAllocationId));
```

Comment:
This particular case handles scenarios where the caller (really only `IpAllocationConstraint`, in this case) provides an `ipAllocationId` where there is no key in the map during scheduling evaluation. The `buildTaskCacheInfo()` populates the map at the beginning of each scheduling interval and puts a `""` as the value if for some reason the Job Descriptor does not have the `ipAllocationId` that was assigned to the Task being examined and thus we cannot determine its zone (not an expected case).

After writing that out, you're right it is confusing. Let me know if that doesn't address your comment.

Reply:
some comments in the code would be helpful about when you'd expect `Optional.of("")` to be the value, it's a bit unclear to me when exactly that would happen. Reading the code I got the impression that the only way a `zoneId` for an `allocationId` on an existing `Task` can not be resolved is when the original `allocationId` has been removed from its containing `Job` (`Job#signedIpAllocations`). Do we allow mutations of ip allocations on existing jobs?

It goes back to my point that not relying on a cache and the current state of the job would be better (i.e.: saving the allocationId + zoneId with the task, as context attributes).

Reply:
and if you think it is still helpful to keep the case where `Optional.of("")` is returned, it is unclear to me why not use `Optional.empty()` instead

Reply:
I added some clarifying comments.

We do not (well, won't) allow mutations of IP allocations on existing jobs (API validations are a later PR).

The zone cache was suggested by Tomasz in his review and added to address his comment. I'm not married to either approach. For now, I changed it the zone cache insert to only insert when there is a valid ID->Zone mapping. Subsequent `taskCache.getZoneIdByIpAllocationId()` calls will result in an `Optional.empty()`. If there' strong opinion about making the IP allocation's zone a task context, I'm happy to do that instead.

Reply:
Thanks for removing the case with `Optional.of("")`. When do you think we would fail to resolve the `zoneId` for a particular `allocationId`? Unless it's a bug, I'm not seeing how it can happen.

If we are only covering a potential corner case with a bug somewhere, it may be worth calling a `CodeInvariants#inconsistent()` so we keep track of it somehow. If I missed something and it is an expected case, some comments about when it happens would be helpful.